keps/sig-multicluster/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,974 @@
+    1. Scale down the source StatefulSet by `1`. Replicas `k` will be terminated (replicas=`k-1`, ordinals.start=`0`)
+       1. This should allow the application to determine that replica `k` is no longer available.
+       2. PDBs associated with each StatefulSet should be adjusted to reflect a reduced availability budget based on `d`
+    2. Move any dependencies of replica `k` to the destination (cluster or namespace)
+       1. This may include namespace resources (PVC, ConfigMap) or cluster scoped resources (PV)
+    3. Scale up the destination StatefulSet by `1`. Replicas `k` will be started (replicas=`N-k`, ordinals.start=`k`)
+       1. The replica `k` should re-advertise its new network identity, through application peer discovery, and network endpoints that reference the existing pod ordinal should be updated.
+ 5. The source StatefulSet should have `0` replicas and destination StatefulSet `N` replicas
+ 6. Clean up the source StatefulSet, and any unused resources safely.
+
+StatefulSets are implicitly numbered starting at ordinal `0`. When pods are being deployed, they are created [sequentially in order](StatefulSet) from pod `0` to pod `N-1`. When pods are being deleted, they are terminated in reverse order from pod `N-1` to pod `0`. This behavior limits the migration scenario where an application operator wants to scale down pods in the source StatefulSet and scale up pods in the destination StatefulSet. If pod `N-1` is removed from the source StatefulSet, there is no mechanism to create only pod `N-1` in a destination StatefulSet without creating pods `[0, N-2]` as well. To do so would lead to the presence of duplicate pod ordinals (eg: pod `0` would exist in both StatefulSets).
+
+Extending StatefulSet to start at an ordinal `k` (eg: `N-1`) would allow the destination StatefulSet to skip over pods `[0, N-2]` when creating pod `N-1`. This allows the original StatefulSet to be sliced at ordinal `k` between the source and destination StatefulSet.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+StatefulSet controller manages pods for a slice of a StatefulSet, within the range `[k, N+k-1]`
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Updating a PDB to safeguard more than one StatefulSet slice
+    *   As StatefulSet slices are scaled up or down, corresponding PDBs can also be adjusted. For example, a PDB corresponding to a slice of `k` replicas could be adjusted to `MinAvailable: k-1` on scale up or down events. Providing guidance and functionality to adjust these PDBs is outside the scope of this KEP.
+* Orchestrating pod movement from one StatefulSet slice to another
+* Managing network connectivity between pods in different StatefulSet slices
+* Orchestrating storage lifecycle of PVCs and PVs across different StatefulSet slices
```

Comment:
Given we have [this mechanism](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1847-autoremove-statefulset-pvcs) for managing statefulset's owned PVC, are you planning to restrict volume delete policy to not use Delete there, given your main goal is migration I'd assume having this there as a reasonable limitation on that other feature.

Reply:
Correct, enabling auto-delete without appropriate safeguards could result in underlying PV data being deleted (say when a source StatefulSet is deleted). If data is referenced across clusters, there isn't any way to enforce this through APIServer. It would be up to the orchestrator performing the migration to handle the safe management of data reference across clusters. I don't think we want to restrict the use of `whenDeleted` or `whenScaled` based on if the user set the `.spec.ordinals.start` field though, since migration is not the only use case.

Reply:
It would be good to point to 1847 in that case when talking about orchestrating storage.